_ndarray.py

```diff
@@ -0,0 +1,103 @@
+import functools
+
+import torch
+
+import _util
+
+NoValue = None
+
+##################### ndarray class ###########################
+
+class ndarray:
+    def __init__(self):
+        self._tensor = torch.Tensor()
+
+    def get(self):
+        return self._tensor
+
+    @property
+    def shape(self):
+        return tuple(self._tensor.shape)
+
+    @property
+    def size(self):
+        return self._tensor.numel()
+
+    @property
+    def ndim(self):
+        return self._tensor.ndim
+
+    @property
+    def dtype(self):
+        return self._tensor.dtype
+
+    @property
+    def strides(self):
+        return self._tensor.stride()   # XXX: byte strides
+
+    # ctors
+    def astype(self, dtype):
+        newt = ndarray()
+        newt._tensor = self._tensor.to(dtype)
+        return newt
+
+    # niceties
+    def __str__(self):
+        return str(self._tensor).replace("tensor", "array_w")
+
+    __repr__ = __str__
+
+
+    ### arithmetic ###
+
+    def __add__(self, other):
+        return self._tensor__add__(other)
+
+    def __iadd__(self, other):
+        return self._tensor.__add__(other)
+
+    def __sub__(self, other):
+        return self._tensor.__sub__(other)
+
+    def __mul__(self, other):
+        return self._tensor.__mul__(other)
+
+    ### methods to match namespace functions
+    def squeeze(self, axis=None):
+        return squeeze(self._tensor, axis)
+
+    def argmax(self, axis=None, out=None, *, keepdims=NoValue):
+        return argmax(self._tensor, axis, out=out, keepdims=keepdims)
+
+    def reshape(self, shape, order='C'):
+        return reshape(self._tensor, shape, order)
+
+
+
+
+def asarray(a, dtype=None, order=None, *, like=None):
+    _util.subok_not_ok(like)
+    if order is not None:
+        raise NotImplementedError
+
+    if isinstance(a, ndarray):
+        return a
+
+    # This and array(...) are the only places which talk to ndarray directly.
+    # The rest goes through asarray (preferred) or array.
+    out = ndarray()
+    out._tensor = torch.asarray(a, dtype=dtype)
+    return out
+
+
+
+def asarray_replacer_1(func):
+    """Asarray the *first* arg, so that the wrapped `func` operates on a
+       torch.Tensor. Then asarray the result.
+    """
+    @functools.wraps(func)
+    def wrapped(x, *args, **kwds):
+        x_tensor = asarray(x).get()
+        return asarray(func(x_tensor, *args, **kwds))
+    return wrapped
```

Comment:
Perhaps we want to wrap any `np.ndarray` that's given to us in the signature? This way this decorator could be written generically for all the functions.

Reply:
If the obs above makes sense, then we could potentially apply the wrapper generically to all the functions defined in the module at the very end of the module, that way we make sure we don't miss any of them and we'd save quite a few lines.

Reply:
Great idea, thanks! I'd prefer to defer it a bit though, to first work through the numpy API patterns (it's... uhm... organically grown).

Reply:
Is decorating a function that returns a Tensor with this replacer actually shorter than not doing so? I'm not yet seeing that yet - only if it has to do more work than wrapping the output arrays. This seems equally short and clearer:
```python
class _ndarray:
    # class name here has to have an underscore probably, so the public ndarray object can use this
    # one but have a signature for __init__ that matches numpy (this is a small detail only).
    def __init__(self, tensor):
        self._tensor = tensor


def absolute(...):
    return _ndarray(result)
```

Reply:
On the other hand, this code block is repeated a lot and seems to need factoring out to be generic across ufuncs:
```python
    _util.subok_not_ok(subok=subok)
    if order != 'K' or casting != 'same_kind' or not where:
        raise NotImplementedError
    if out is not None:
      # XXX dtypes, casting
        out = out.to(dtype)
```

Reply:
Note that the decorator aso wraps the first argument (and potentially others). Also, as discussed above, the decorator could potentially be applied to all functions in the module a posteriori so that we don't miss any of them. This would very much reduce the amount of boilerplate needed all across.

Reply:
> Note that the decorator aso wraps the first argument (and potentially others).

It's not clear to me that it has to do that, or that it's better. The current implementation also explicitly does `torch.as_tensor(x)` internally anyway (maybe a leftover).

The problem with these kinds of decorators, unless they do more than wrap inputs and outputs, is that they're hard to read for no real gain. Right now, `out` isn't handled for example - and that can also be an array. There will be lots of one-offs where some other argument can be an array, and there's no decorator for exactly that thing. So now you have to go write yet another decorator with a slight twist, or mix decorator with manual wrapping.




> This would very much reduce the amount of boilerplate needed all across.

I think it makes sense to figure out if this is actually true. Here is how I'd write `absolute`, and I like how it looks better than with the decorator approach:
```python

def handle_ufunc_kwargs(subok=None, order=None, out=None, dtype=None):
    # This is mostly factoring out what was already written, it's not complete
    _util.subok_not_ok(subok=subok)
    if order != 'K' or casting != 'same_kind' or not where:
        raise NotImplementedError

    if out is not None:
        out = as_tensor(out)  # unwrap
        if dtype is not None:
            if dtype != out.dtype:
                raise ValueError("given `out` array doesn't have the right dtype")

    return out


def absolute(x, /, out=None, *, where=True, casting='same_kind', order='K',
          dtype=None, subok=False, **kwds):
    x = as_tensor(x)
    out = handle_ufunc_kwargs(subok=subok, order=order, out=out, dtype=dtype)

    result = torch.absolute(x, out=out)
    if dtype is not None:
        result = result.to(dtype)
    
    return _ndarray(result)
```

Reply:
The ufunc code block is actually autogenerated from template, https://github.com/Quansight-Labs/numpy_pytorch_interop/blob/ndarrays_at_interface/autogen/gen_ufuncs.py#L72 
and it's of course a stub waiting to handle dtypes and out arrays (shapes, types etc), also casting. 

Given how this develops for now, I'd like to push the current approach for a bit more to see various patterns in numpy. E.g., I'd rather not detail the ufunc handler yet, because e.g. `subok` is used in both ufuncs and non-ufuncs. Same for generalizing the decorators or other approaches, let's first see what is needed, then clean up and consolidate. I am not yet certain that what is here now is truly representative, need to convert some more functions.

Whether a single decorator can be applied to all args, I suspect not because array_like treatment differs by function:
- func(array_like, array_like) -- e.g. binary ufuncs
- func(array_like, tuple) -- e.g. `np.reshape(array, new_shape)`
- func(*array_likes) -- e.g. `atleast_1d` 

One other thing I'd rather keep rolling with is not using `ndarray` class (with or with or without a leading underscore) for wrapping the outputs. The public `ndarray` class will need to have numpy-compatible signature. So for now I'm using `asarray` as a central thing which knows about the internals about the `ndarray` class. Let me see how far this takes us.

Reply:
> One other thing I'd rather keep rolling with is not using `ndarray` class (with or with or without a leading underscore) for wrapping the outputs.

That's fine to defer I'd say, it's easy to tweak later and it's a ~10 characters change.

Reply:
> x = as_tensor(x)

BTW, this does not handle ndarrays.